Add rpc requests article (#506)
enviodev/docs: docs/HyperIndex/Guides/rpc-requests.md

---
id: rpc-requests
title: RPC Requests
sidebar_label: RPC Requests
slug: /rpc-requests
---

# RPC Requests guide

> TLDR; The repo for the code base can be found [here](todo_add_link)

Reviewer: Let's set it to this, and when we transfer the repo we can make sure it matches the below :)
```suggestion
TLDR; The repo for the code base can be found [here](https://github.com/enviodev/rpc-token-data-example)
```

**Problem**  
At first glance this may appear to be a simple task as we are only indexing over one event, the [pool created](https://docs.uniswap.org/contracts/v3/reference/core/interfaces/IUniswapV3Factory#poolcreated) event. However, there is one catch, have a look at the event signature: 

```yaml
PoolCreated(address indexed token0, address indexed token1, uint24 indexed fee, int24 tickSpacing, address pool)
```

The event only provides the token addresses, not the token name, symbol, and decimals. So how are we to fetch this extra information in our event handler? 

**Solution**  
In order to do this, we need to make RPC requests to the token contract.

### Prerequisites

This guide assumes basic familiarity with the viem library for making contract calls. Check out the [viem documentation](https://viem.sh/) for more information. This [medium article](https://medium.com/@0xape/typescript-and-viem-quickstart-for-blockchain-scripting-3f1846970b6f) is highly recommended for a gentle introduction to viem with a very similar example to what we are doing here.

### Part 1: Create our Uniswap V3 `poolcreated` indexer

`npx envio init`

> Contract address: `0x1F98431c8aD98523631AE4a59f267346ea31F984`

Reviewer: At this point it's not entirely apparent which network this is. 
```suggestion
Ethereum mainnet contract address: `0x1F98431c8aD98523631AE4a59f267346ea31F984`
```
Could also simply hyperlink the address to an explorer - here's a pretty cool explorer built on top of envio
https://scope.sh/1/address/0x1F98431c8aD98523631AE4a59f267346ea31F984

> config.yaml

:::info
We remove the `FeeAmountEnabled` and `OwnerChanged` events as they are not relevant to our use case.
:::

```yaml
name: uniswap-v3-factory-token-indexer
networks:
- id: 1
  start_block: 12369620
  contracts:
  - name: UniswapV3Factory
    address:
    - 0x1F98431c8aD98523631AE4a59f267346ea31F984
    handler: src/EventHandlers.ts
    events:
    - event: PoolCreated(address indexed token0, address indexed token1, uint24 indexed fee, int24 tickSpacing, address pool)
rollback_on_reorg: false
```

> shema.graphql

:::info
Simplify our schema to only include the pool and token information.
:::

```graphql
type Token {
  id: ID! #address
  name: String!
  symbol: String!
  decimals: Int!
}

type Pool {
  id: ID! # address
  token0: Token! 
  token1: Token!
  fee: BigInt!
  tickSpacing: BigInt!
  pool: String!
}
```

> src/EventHandlers.ts

```typescript
import {
  UniswapV3Factory,
  Pool,
} from "generated";
import { getTokenDetails } from "./tokenDetails";

UniswapV3Factory.PoolCreated.handler(async ({ event, context }) => {
  const entity: Pool = {

>     id: `${event.chainId}_${event.block.number}_${event.logIndex}`,

Reviewer: I usually use dashes instead of underscores, both are perfectly fine just might be confusing for some users so I think in general we should try stick to one. I couldn't find anything now in the docs with one or the other but I know all the examples I've done would have '-'
```suggestion
    id: `${event.chainId}-${event.block.number}-${event.logIndex}`,
```

    token0_id: event.params.token0,
    token1_id: event.params.token1,
    fee: event.params.fee,
    tickSpacing: event.params.tickSpacing,
    pool: event.params.pool,
  };
  context.Pool.set(entity)

  try {
    const {name: name0, symbol: symbol0, decimals: decimals0} = await getTokenDetails(event.params.token0, event.chainId);
    context.Token.set({
      id: event.params.token0,
      name: name0,
      symbol: symbol0,
      decimals: decimals0
    })
  } catch (error) {
    console.log(error)
    return
  }

  try {

>     const {name: name1, symbol: symbol1, decimals: decimals1} = await getTokenDetails(event.params.token0, event.chainId);

Reviewer: I think this should be a 1?
```suggestion
    const {name: name1, symbol: symbol1, decimals: decimals1} = await getTokenDetails(event.params.token1, event.chainId);
```

    context.Token.set({
      id: event.params.token1,
      name: name1,
      symbol: symbol1,
      decimals: decimals1,
    })
  } catch (error) {
    console.log(error)
    return
  }
});

```

### Part 2: Fetch token information

Note how we use the `multicall` feature from Viem to batch our 3 queries into one request to improve performance and help avoid rate limiting. Very useful!

> Here, we use an alchemy API key to make RPC requests to the ethereum mainnet. You can get an alchemy API key [here](https://www.alchemy.com/) and add it to your `.env` file. Alternatively, use any other rpc provider of your choice.

Reviewer: Alchemy is definitely a good robust option, might be worth linking to chainlist.org for finding some alternative free rpc providers

Reviewer: In fact since we partner with dRPC could you rather mention them :)

> src/tokenDetails.ts

```typescript
import { createPublicClient, getContract, http } from 'viem'
import { mainnet } from 'viem/chains'
import { Cache, CacheCategory } from "./cache";
import { ERC20ABI } from './constants';

> const apiKey = process.env.ALCHEMY_API_KEY;

Reviewer: I suggest rather making this the full rpc endpoint than an api key as it's more flexible to switch between providers. ie something like
```suggestion
const apiKey = process.env.RPC_URL;
```
